File: leetcode/040_combinationsumII.py
class Solution(object):
    def backtrack(self, currlist, start, nums, target, result):
        if sum(currlist) == target:
            result.append(currlist)
        if sum(currlist) > target:
            return
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i-1]:
                continue
            self.backtrack(currlist + [nums[i]], i+1, nums, target, result)
    
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates = sorted(candidates)
        result = []
        self.backtrack([],0,candidates,target,result)
        return results

File: leetcode/040_combinationsumII.ts
function combinationSum2(candidates: number[], target: number): number[][] {
    if (candidates.length === 0) return []
    candidates.sort()
    const res = []
    generate(candidates, target, 0, [], 0, res)
    return res
};

function generate(candidates: number[], target: number, currSum: number, currList: number[], idx: number, results: number[][]) {
    if (currSum > target) return
    if (currSum === target) {
        results.push(currList)
        return
    }

    for (let i = idx; i < candidates.length; i++) {
        if (i > idx && candidates[i] === candidates[i-1]) continue
        generate(candidates, target, currSum + candidates[i], [...currList, candidates[i]], i + 1, results)
    }
}
